src/adapter/src/catalog/migrate.rs

```diff
@@ -163,6 +165,312 @@ pub(crate) async fn migrate(
+    };
+    use mz_storage_types::sources::postgres::{
+        ProtoPostgresSourceExportStatementDetails, ProtoPostgresSourcePublicationDetails,
+    };
+    use mz_storage_types::sources::proto_source_export_statement_details;
+    use mz_storage_types::sources::ProtoSourceExportStatementDetails;
+    use prost::Message;
+
+    struct Rewriter<'a, 'b> {
+        tx: &'a Transaction<'b>,
+    }
+
+    impl<'ast> VisitMut<'ast, Raw> for Rewriter<'_, '_> {
+        fn visit_create_subsource_statement_mut(
+            &mut self,
+            node: &'ast mut CreateSubsourceStatement<Raw>,
+        ) {
+            match &node.of_source {
+                // Not a source export subsource
+                None => (),
+                Some(source) => {
+                    let details = node
+                        .with_options
+                        .iter()
+                        .find(|o| o.name == CreateSubsourceOptionName::Details);
+                    if details.is_some() {
+                        // this subsource has already had its details written, so it's either new or was already
+                        // migrated
+                        return;
+                    }
+
+                    let external_reference = node
+                        .with_options
+                        .iter()
+                        .find(|o| o.name == CreateSubsourceOptionName::ExternalReference)
+                        .expect("subsources must have external reference");
+                    let (external_schema, external_table) = match &external_reference.value {
+                        Some(WithOptionValue::UnresolvedItemName(name)) => {
+                            let name_len = name.0.len();
+                            (
+                                name.0[name_len - 2].clone().into_string(),
+                                name.0[name_len - 1].clone().into_string(),
+                            )
+                        }
+                        _ => unreachable!("external reference must be an unresolved item name"),
+                    };
+
+                    let source = match &source {
+                        RawItemName::Name(_) => {
+                            panic!("named item reference in durable catalog")
+                        }
+                        RawItemName::Id(id, _) => {
+                            let gid = id
+                                .parse()
+                                .expect("RawItenName::Id must be uncorrupted GlobalId");
+                            self.tx.get_item(&gid).expect("source must exist")
+                        }
+                    };
+                    let source_statement =
+                        mz_sql_parser::parser::parse_statements(&source.create_sql)
+                            .expect("parsing persisted create_sql must succeed")
+                            .into_element()
+                            .ast;
```

Comment:
If we add a source AST migration, then this might fail if it's run before the source AST migrations. It's fine for now, but let's keep this in the back of our minds in case something breaks this release.